### ISSUE TITLE
DCC Certificate request in App says "Error" while not been processed from DCC backend

### ISSUE THREAD

**Issue by mbay-ODW** [corona-warn-app/cwa-app-ios#4150]
## Avoid duplicates
* [ X] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [ ] Bug is specific for iOS only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [ X] Bug is not already reported in another issue

## Technical details

- Device name: iPhone 13 mini
- iOS version: 15.3
- App version: 2.16.2

## Describe the bug

We just implemented the DCC validation and find one of the tables disturbing. While the customer within the app requested the certificate and the DCC certificate is not yet validated a message appears that says error. This caused multiple questions from customer side.

<img width="388" alt="image" src="https://user-images.githubusercontent.com/58465677/151692124-f85ddd93-66c9-49ad-8468-43d20baf0236.png">

From a definition point that is not an error but a pending request, which should be labeled like that. In the implementation guidelines of DCC is no fixed request validation time given such that any time would fit those (e.g. every 10 minutes).

## Steps to reproduce the issue

1. Customer scans CWA Code that has dgb=true
2. Customer requests DCC certificate
3. While DCC request is pending error is shown

## Expected behaviour

While the DCC request is pending, the message should say pending, not error. There is no hint in the message that this could be due to the fact that the request has not been processed.

## Additional context

---
Internal Tracking ID: [EXPOSUREAPP-13223](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-13223)

**Comment by thomasaugsten** [comment]
I think you referring to the hint
"Ihr Zertifikat liegt noch nicht vor. Bitte versuchen Sie es noch einmal. Sollte der Fehler weiterhin bestehen, kontaktieren Sie bitte die technische Hotline über App-Informationen -> Technische Hotline." Code 202

Normally this hint should not shown to the user because after a negative test result the app request the DCC from the dcc server in most cases this DCC should be already available. If not the dcc server will sign the dcc from test provider and provides the dcc in 2-10 seconds.

The app is requesting the dcc after receiving a negativ dcc. If its not available not hint to the user is shown, the app will wait 10 seconds and retry to receive the dcc. If the second try is not available the user receive the hint with code 202.

In this cases the test provider is not fetching the public keys from the dcc provider properly or is not submitting the dcc with  the negative test result.

**Comment by mbay-SAG** [comment]
Thanks for the fast reply. In deed the App shows for a couple of seconds the message, that the certificate will be generated and this could take a couple of minutes. After around 10s the above mentioned error warning appear.

The DCC implementation does not force a certain process interval for validating those requests. This is why we chosed to process those requests every 5 minutes which lead to the above mentioned behavior. Since the app says "This could take up to several minutes" I still think that the Error after 10s is confusing, it should stay within the pending state without error.

**Comment by thomasaugsten** [comment]
Our implementation information was that the process interval to pull the public keys is up to the provider but the dcc has to be send with the negative test result and the processing on the dcc server takes not more than 10seconds.
The phrase minutes was selected to include network problems on user side or peak load situation on the dcc server.

Please double check with the dcc server team when the dcc has to be submitted with the negative test result or x minutes after the negative test result.

**Comment by mbay-SAG** [comment]
The implementation process involves a manuel process on customer side within the app. CWA test result and DCC are decoupled processes only coupled via the request within the CWA.

I´m refering to the Flow taken from: https://github.com/corona-warn-app/cwa-quicktest-onboarding/blob/master/DCC_Creation_Flow.png

![DCC_Creation_Flow](https://user-images.githubusercontent.com/58465677/151707463-9145d2bc-cba0-41f5-ba9b-b75278a82315.png)

Customer requests within the App, that he needs an DCC (Step 4). Therefore the Partner Packend can not send the result (step 1)  to the CWA Backend together with the DCC (step 7).

After requesting the DCC within the App manually via click, the polling on DCC Requests (Step 5) will show the particular testID. After that the Partner backend can provide the DCC (Step 7). Therefore the Error Message appears since we are not using a very high polling interval for the DCC requests.

**Comment by thomasaugsten** [comment]
The flow chart is not precise. The request dcc and the first test result poll happens at the same time because the user request the dcc when he scan the QR code before he receive the test result. This means step 5 should happen for the most cases before submitting the test result. Except the rare cases when the test result is immediately available after scanning. 

But please check with the dcc server Team for best practices.

**Comment by mbay-SAG** [comment]
That is Not True. The dcc has to be requested manually within the App. There is no Direct dcc request while scanning the qr Code. There is an additional click requiriert. This is described in the official DCC backend Description. How can The linked architecture be "Not precise" if its part of the acceptance Test and specification. 

In [Link](https://github.com/corona-warn-app/cwa-quicktest-onboarding/blob/master/docs/DCC_Data_Model_and_Communications.pdf) there is Said:

![image](https://user-images.githubusercontent.com/58465677/151710956-41f003b6-9218-4cf4-a8fa-4b3b10a147d9.png)

See also:
https://github.com/corona-warn-app/cwa-quicktest-onboarding/wiki/Anbindung-an-CWA-mit-Verwendung-von-DCCs#backend-integration

**Comment by Ein-Tim** [comment]
@mbay-SAG 

@thomasaugsten Means this. After you scan in a test QR code, you are asked to decide whether you want a test DCC or not:

https://user-images.githubusercontent.com/67682506/151712134-7132febd-39ed-43bb-b0b1-a3f9432baa59.MP4

**Comment by thomasaugsten** [comment]
Sorry there was a misunderstanding on my side. The User give the dcc consent during scanning before polling the test result but will submit the public after the app is receiving the negative test result.
The user have to wait for your next poll interval. 

What is your exact request to improve the hint for the user
![image](https://user-images.githubusercontent.com/6367738/151712951-20b7aa5d-6043-494f-ab09-3c604b1cb5db.png)

**Comment by mbay-SAG** [comment]
Thanks @Ein-Tim for the recording. The improvement would be on that screenshot:

<img width="388" alt="151692124-f85ddd93-66c9-49ad-8468-43d20baf0236" src="https://user-images.githubusercontent.com/58465677/151713080-148a1c85-657d-4943-878e-054326230e8c.png">

App:

1. It says that there is an ERROR/FEHLER, which is basically missleading. True is, that the request is pending in the backend and waits for validation. This can take a couple of minutes.

Backend:

1. On top of that I would recommend to get in contact with the DCC Backend Team to give guidelines on the expected poll frequency for backend partners, currently there is none. Personally we found an interval of every 2-5 minutes more than sufficient in terms of performance.

**Comment by thomasaugsten** [comment]
We will create a bug ticket for this
For the backend topic please contact the DCC Team

**Comment by Ein-Tim** [comment]
This is what I just saw. I received the test result in the app, switched to the certificates tab and saw this. After ca. 5 sec. the DCC was there and everything was fine. This is how it should be, shouldn't it?

<img src="https://user-images.githubusercontent.com/67682506/151714105-a3658a9f-5476-46ce-bdc8-9cf4244337e8.jpeg" width="300">

**Comment by mbay-SAG** [comment]
> For the backend topic please contact the DCC Team

How can I contact That Team?

**Comment by mbay-SAG** [comment]
> This is what I just saw. I received the test result in the app, switched to the certificates tab and saw this. After ca. 5 sec. the DCC was there and everything was fine. This is how it should be, shouldn't it?
> 
> 
> 
> <img src="https://user-images.githubusercontent.com/67682506/151714105-a3658a9f-5476-46ce-bdc8-9cf4244337e8.jpeg" width="300">
> 
> 

Thats true if the Partner backend polls for dcc requests in a high frequency (e.g. 5s) However this is actually not really necessary nor recommended in the specifications. In terms of Performance this Would create a Lot of Workload on both Servers.

**Comment by Ein-Tim** [comment]
I never experienced something else so I highly recommend that you cross check with other partners what their implementations are. 😅

In this case the test analyzed by "Labor Becker".

**Comment by mbay-SAG** [comment]
I think I realized now why it comes to that screen, thanks for your active communication so far. 

We have an implementation where the QRCode is delivered only AFTER the result is already available. This is due to many reasons, e.g. minimize customer/worker contacts and the fact that we use both, RAT and non RAT certified Tests. (RAT certified Tests use a different QR code than non RAT Tests due to the dbg field).

So if the customer scans the Code AFTER the result is already delivered, the mentioned ERROR is shown since the DCC polling job did not yet get the DCC request within the fixed 10s time frame of the App but the result is already there. I double checked with the CWA Implementation guidelines, there is no recommendation to NEED provide the CWA BEFORE the result is set.

How can I contact the DCC Team as suggested by @thomasaugsten?

**Comment by thomasaugsten** [comment]
I think cwa-schnelltest-onboarding@t-systems.com can help or forward the request

**Comment by mbay-SAG** [comment]
Thanks @ndegendogo, I totally agree that the QRCode delivery strongly depends on the workflow and thus can not be described as "rare cases".

**Comment by mbay-SAG** [comment]
@Ein-Tim I have no Android available, thus can not test.

**Comment by Ein-Tim** [comment]
@thomasaugsten Will there something be changed reg. the behavior of the app? If not, I suggest to close this issue as a `wont fix`.

**Comment by MaxPeal** [comment]
close this issue as a wont fix, is in my opinion a bit problematic,
as the UI / Error Text is misleading,
a better Error Text with more info would be relay helpful.

**Comment by mbay-ODW** [comment]
What I actually dislike is that there is no Communication here.

**Comment by larswmh** [comment]
@mbay-ODW we appreciate your patience and will keep you updated once we have information to share. 

---
Internal Tracking ID: [EXPOSUREAPP-13223](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-13223)

**Comment by larswmh** [comment]
Closing this ticket as its related internal ticket was set to obsolete.

> At this point the certificate is expected to be available, but it is not. Therefore it's an error. The CWA does not know if it's just pending and will arrive later or if it will not arrive at all due an error in the process.
